### PR TITLE
Refactor `ModDifficultyAdjust` to more elegantly track user override status

### LINKED ISSUE #13799: Adjust difficulty mod settings reset issue
**Describe the bug:** 
In multiplayer, when you set the mod a first time and play a map, it works as intended, however, the settings are reset on the next map chosen, it basically acts like the mod isn't enabled, this doesn't happen in solo.

In solo, when you play a map matching the settings of the mod, those settings are unlocked, on the next map you play it will be the settings of the map and not the mod:

> Adjust difficulty: AR10
> Play a map with AR9
> Change is applied
> Play a map with AR10
> Well nothing happens it matches
> Play any map, it won't be AR10 anymore

**osu!lazer version:** 
2021.703.0

### PR DESCRIPTION
**Pull request by peppy**
As discussed in discord. I'll let the code speak for itself here, not much to add.

- Supersedes https://github.com/ppy/osu/pull/13814 (includes it but adjusts one or two expectations)
- Fixes https://github.com/ppy/osu/issues/13799

https://user-images.githubusercontent.com/191335/124893209-7feb6380-e015-11eb-844c-226257919407.mov

### CODE REVIEW

File: osu.Game.Tests/Visual/UserInterface/TestSceneModDifficultyAdjustSettings.cs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+            setSliderValue("Circle Size", 9);
+            checkSliderAtValue("Circle Size", 9);
+            checkBindableAtValue("Circle Size", 9);
+
+            setBeatmapWithDifficultyParameters(4);
+
+            checkSliderAtValue("Circle Size", 9);
+            checkBindableAtValue("Circle Size", 9);
+        }
```
Comment (bdach) [inline osu.Game.Tests/Visual/UserInterface/TestSceneModDifficultyAdjustSettings.cs:155]: Bit of a discrepancy between the test name here and the test scenario. I *assume* `setSliderValue()` should have been called with 9 (or the second `setBeatmapWithDifficultyParameters()` with 4)?

**Comment by bdach** [comment]
8 out of 9 files in this diff I like the look of a lot. The removal of these old hacks from `ModDifficultyAdjustment` is a sight for sore eyes.

The one I'm not super sure of is `DifficultyAdjustSettingsControl`. It starts to make sense after a bit of staring, but it's still a little complex (and maybe inevitably so, given that what it's doing is a little complex on its own). Maybe it can be improved with some renaming, not sure. I would definitely appreciate a second pair of eyes to look at it to make sure it's not just me.

File: osu.Game.Tests/Visual/UserInterface/TestSceneModDifficultyAdjustSettings.cs (hunk 1 above)
Comment (peppy) [reply osu.Game.Tests/Visual/UserInterface/TestSceneModDifficultyAdjustSettings.cs:155]: Yep, good catch.

**Comment by peppy** [comment]
Definitely agree about that class being on the verge of too complicated. I'll probably do another pass myself tomorrow but I'm not sure how much it can be improved, as it is both fitting the existing `SettingsItem` mould and adding one extra dimension.

Even if it can't be heavily improved, at least it's now confined to this one class, though.

**Comment by peppy** [comment]
@bdach give that another read now. i think the class should be a lot easier to mentally parse.

**Review by bdach — DISMISSED** [review]
Yeah I like this as is. Not sure whether another pair of eyes is needed.

**Review by frenzibyte — CHANGES_REQUESTED** [review]
Been randomly playing with the controls and noticed this:

https://user-images.githubusercontent.com/22781491/125178368-f2626c00-e1ec-11eb-8b52-d11fb039488a.mp4

Seems reproducible by first selecting then deslecting the mod, then activate again and change the values (w/o extended limits checked I think)

---

Also missing `SettingControlType` specification for the osu!catch variant, as "suggested" in the comments below.

**Comment by peppy** [comment]
What is the video showing? I've watched 5 times and still have no idea.

**Comment by frenzibyte** [comment]
"Approach rate" setting no longer showing the restore value indicator.

**Comment by peppy** [comment]
Thanks. I remember there being some unbind code somewhere that may still be being triggered.

**Comment by bdach** [comment]
I believe I've resolved the issue. As far as I can tell this was caused by storing the `DifficultyBindable` raw in drawables (`DifficultyAdjustSettingsControl` and its inner `SliderControl`), which in turn caused the framework-side automatic unbinding code to fire and unbind the mod's `DifficultyBindable`s (as they were the same reference as the controls').

The fix is to add a specialised implementation of `GetBoundCopy()`. It's a bit dodgy around `MaxValue` and `ExtendedMaxValue` (assumes they're essentially constants); they can be changed to bindables if need be, but I figured the bare minimum might be enough for now.

**Review by frenzibyte — DISMISSED** [review]

File: osu.Game/Rulesets/Mods/DifficultyBindable.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        /// <summary>
+        /// Whether the extended limits should be applied to this bindable.
+        /// </summary>
+        public readonly BindableBool ExtendedLimits = new BindableBool();
+
+        /// <summary>
+        /// An internal numeric bindable to hold and propagate min/max/precision.
+        /// The value of this bindable should not be set.
+        /// </summary>
+        public readonly BindableFloat CurrentNumber = new BindableFloat
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Mods/DifficultyBindable.cs:21]: If this bindable should not be set, then it should probably be `internal`?

**Review by smoogipoo — DISMISSED** [review]
Looks okay

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**